render_silhouette_regmarks: render rego mark (From miLORD1337)

mofosyne opened the pull request:
Found [miLORD1337](https://github.com/miLORD1337) post in https://github.com/fablabnbg/inkscape-silhouette/issues/79 where he was proposing his work.

His code is GPLed and looks suitable for integration into this extension for convenience. For user experience, I've made the width and length parameter optional so that the user only needs to care about the spacing from edge of the document (we can generally assume that the document would be the print size (e.g. A4)... if that's not the case then the user can still override that value anyway.

Next thing we could do is update the reg mark page in the main script to also make these same parameter optional as well.

https://github.com/miLORD1337/silhouette-regmarks

![image](https://github.com/fablabnbg/inkscape-silhouette/assets/827793/0881b72a-027f-44a3-8277-db907f06ae19)

mofosyne committed:
render rego mark. Copied from miLORD1337 silhouette-regmarks and adjusted to make width and length param optional

mofosyne committed:
reg line should be 0.3mm for better accuracy

mofosyne committed:
update readme on reg mark rendering

mofosyne committed:
keepout area marker in regmarker render

mofosyne committed:
log message adjustment

mofosyne committed:
render layer named cahnged from silhouette-regmark to Regmarks for consistency purpose

mofosyne committed:
regmark render: added self.svg.unittouu('1mm') to scale based on document property

t0b3 commented on render_silhouette_regmarks.inx:
```diff
@@ -0,0 +1,26 @@
+  <effect>
+    <object-type>all</object-type>
+    <effects-menu>
+     <submenu _name="Render">
```
_name is deprecated. use `name` instead.

t0b3 commented on render_silhouette_regmarks.inx:
```diff
@@ -0,0 +1,26 @@
+    </effects-menu>
+  </effect>
+  <script>
+    <command reldir="extensions" interpreter="python">render_silhouette_regmarks.py</command>
```
`location="inx"` is what is used for better portability.

t0b3 commented on render_silhouette_regmarks.inx:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<inkscape-extension xmlns="http://www.inkscape.org/namespace/inkscape/extension">
```
`translationdomain="inkscape-silhouette"` will enable translatability

t0b3 commented on render_silhouette_regmarks.inx:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<inkscape-extension xmlns="http://www.inkscape.org/namespace/inkscape/extension">
+  <_name>Classic</_name>
```
_name is deprecated. use `name`  instead.
you may replace `Classic` with your proposal for a good name...

t0b3 commented on render_silhouette_regmarks.inx:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<inkscape-extension xmlns="http://www.inkscape.org/namespace/inkscape/extension">
+  <_name>Classic</_name>
+  <id>org.inkscape.render.silhouette-regmarks</id>
```
for consistency you may use a prefix with `com.github.fablabnbg.inskscape-silhouette.silhouette_...`

t0b3 commented on render_silhouette_regmarks.inx:
```diff
@@ -0,0 +1,26 @@
+<inkscape-extension xmlns="http://www.inkscape.org/namespace/inkscape/extension">
+  <_name>Classic</_name>
+  <id>org.inkscape.render.silhouette-regmarks</id>
+  <!--<dependency type="executable" location="extensions">inkex.py</dependency>-->
```
Obsolete. You may drop this line.

t0b3 commented on render_silhouette_regmarks.inx:
```diff
@@ -0,0 +1,26 @@
+  <_name>Classic</_name>
+  <id>org.inkscape.render.silhouette-regmarks</id>
+  <!--<dependency type="executable" location="extensions">inkex.py</dependency>-->
+  <dependency type="executable" location="extensions">render_silhouette_regmarks.py</dependency>
```
`location="inx"` improves portability

t0b3 commented on render_silhouette_regmarks.inx:
```diff
@@ -0,0 +1,26 @@
+  <id>org.inkscape.render.silhouette-regmarks</id>
+  <!--<dependency type="executable" location="extensions">inkex.py</dependency>-->
+  <dependency type="executable" location="extensions">render_silhouette_regmarks.py</dependency>
+  <param name="regoriginx" type="float" min="10.0" max="10000" _gui-text="Position of regmark from document left [mm]">10</param>
```
_gui-text is deprecated. use `gui-text` instead.

t0b3 commented on render_silhouette_regmarks.inx:
```diff
@@ -0,0 +1,26 @@
+  <!--<dependency type="executable" location="extensions">inkex.py</dependency>-->
+  <dependency type="executable" location="extensions">render_silhouette_regmarks.py</dependency>
+  <param name="regoriginx" type="float" min="10.0" max="10000" _gui-text="Position of regmark from document left [mm]">10</param>
+  <param name="regoriginy" type="float" min="10.0" max="10000" _gui-text="Position of regmark from document top [mm]">10</param>
```
_gui-text is deprecated. use `gui-text` instead.

t0b3 commented on render_silhouette_regmarks.inx:
```diff
@@ -0,0 +1,26 @@
+  <dependency type="executable" location="extensions">render_silhouette_regmarks.py</dependency>
+  <param name="regoriginx" type="float" min="10.0" max="10000" _gui-text="Position of regmark from document left [mm]">10</param>
+  <param name="regoriginy" type="float" min="10.0" max="10000" _gui-text="Position of regmark from document top [mm]">10</param>
+  <param indent="2" name="regdistance_help" type="description">Spacing of the registration mark edges</param>
```
param type=description is deprecated. use `<label> w/o name or type instead.

t0b3 commented on render_silhouette_regmarks.inx:
```diff
@@ -0,0 +1,26 @@
+  <param name="regoriginx" type="float" min="10.0" max="10000" _gui-text="Position of regmark from document left [mm]">10</param>
+  <param name="regoriginy" type="float" min="10.0" max="10000" _gui-text="Position of regmark from document top [mm]">10</param>
+  <param indent="2" name="regdistance_help" type="description">Spacing of the registration mark edges</param>
+  <param name="regwidth" type="float" min="0.0" max="10000" _gui-text="X mark distance [mm]"></param>
```
_gui-text is deprecated. use `gui-text` instead.

t0b3 commented on render_silhouette_regmarks.inx:
```diff
@@ -0,0 +1,26 @@
+  <param name="regoriginy" type="float" min="10.0" max="10000" _gui-text="Position of regmark from document top [mm]">10</param>
+  <param indent="2" name="regdistance_help" type="description">Spacing of the registration mark edges</param>
+  <param name="regwidth" type="float" min="0.0" max="10000" _gui-text="X mark distance [mm]"></param>
+  <param name="reglength" type="float" min="0.0" max="10000" _gui-text="Y mark distance [mm]"></param>
```
_gui-text is deprecated. use `gui-text` instead.

t0b3 commented on render_silhouette_regmarks.inx:
```diff
@@ -0,0 +1,26 @@
+  <param indent="2" name="regdistance_help" type="description">Spacing of the registration mark edges</param>
+  <param name="regwidth" type="float" min="0.0" max="10000" _gui-text="X mark distance [mm]"></param>
+  <param name="reglength" type="float" min="0.0" max="10000" _gui-text="Y mark distance [mm]"></param>
+  <param indent="2" name="regdistance_help" type="description">Distance between registration mark edges</param>
```
param type=description is deprecated. use ` w/o name or type instead.

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,168 @@
+				type = float, dest = "regoriginx", default = 15.0, help="X mark origin from left [mm]")
+		self.arg_parser.add_argument("--rego-y", "--regoriginy", 
+				type = float, dest = "regoriginy", default = 20.0, help="X mark origin from top [mm]")
+		self.arg_parser.add_argument("--verbose", dest = "verbose", type = inkex.Boolean, default = False, help="enable log messages")
```
you can pack this section into it's own function `add_arguments()`
```
    def add_arguments(self, pars):
            pars.add_argument(...)
            etc...
```

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,168 @@
+	#SVG rect element generation routine
+	def drawRect(self, size, pos, name):
+		x, y = [pos * self.svg.unittouu('1mm') for pos in pos  ]
+		w, h = [pos * self.svg.unittouu('1mm') for pos in size ]
```
`unittouu` is discouraged.
see https://gitlab.com/inkscape/extensions/-/blob/master/inkex/elements/_base.py#L590-595

mofosyne replied:
Attempted... but didn't seem to come out right

```
		x, y = [self.svg.unit_to_viewport(pos, 'mm') for pos in pos  ]
		w, h = [self.svg.unit_to_viewport(pos, 'mm') for pos in size ]
```

t0b3 replied:
what was the issue?

mofosyne replied:
When I did something like this

```python
	#SVG rect element generation routine
	def drawRect(self, size, pos, name):
		x, y = [self.svg.unit_to_viewport(pos, 'mm') for pos in pos  ]
		w, h = [self.svg.unit_to_viewport(pos, 'mm') for pos in size ]
		return inkex.Rectangle.new(x, y, w, h, id=name, style='fill: black;')
		
	#SVG line element generation routine
	def drawLine(self, posStart, posEnd, name):
		x1, y1, = [self.svg.unit_to_viewport(pos, 'mm') for pos in posStart]
		x2, y2, = [self.svg.unit_to_viewport(pos, 'mm') for pos in posEnd  ]
		line_style = 'stroke: black; stroke-width: '+str(self.REG_MARK_LINE_WIDTH_MM * self.svg.unittouu('1mm'))+';'
		return inkex.Line.new((x1, y1), (x2, y2), id=name, style=line_style)
```

It renders okay on a new file, but on the existing template that has different document scaling the rendered mark is smaller than expected when compared against the current regmark as shown in the screenshot below

![image](https://github.com/fablabnbg/inkscape-silhouette/assets/827793/ee9131ac-ad9f-4219-af16-e70a7f14b21e)

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,168 @@
+		rect.set('id', name)
+		rect.set('width', str(w))
+		rect.set('height', str(h))
+		rect.set('style', 'fill: black;')
```
you may simply use 
```
inkex.Rectangle(x, y, width, height, ...)
```
where you can add any attrs for ... (i.e. name, style, etc.)
see https://inkscape.gitlab.io/extensions/documentation/source/inkex.elements._polygons.html#inkex.elements._polygons.Rectangle

So probably you won't need this function at all but call `Rectangle()` directly...

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,168 @@
+		# > The registration mark thickness is actually very important. For some reason, 0.3 mm marks work perfectly. 
+		# > The thicker you get, the less accurate registration will be. ~~~ galaxyman47
+		REG_MARK_LINE_WIDTH_MM = 0.3
+		line.set('style', 'stroke: black; stroke-width: '+str(REG_MARK_LINE_WIDTH_MM* self.svg.unittouu('1mm'))+';')
```
same here. You may call  `inkex.Line()` directly and drop this function entirely...
see https://inkscape.gitlab.io/extensions/documentation/source/inkex.elements._polygons.html#inkex.elements._polygons.Line

... and consider lines move according to their width. alternatively you try (thin) rectangles to keep one side perfectly aligned.

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,168 @@
+		REG_LINE_MM = 20
+		REG_KEEPOUT_MM = 2
+
+		svg = self.document.getroot()
```
this line can be dropped. it's handled in inkex already.

mofosyne replied:
unsure how to replace `etree.SubElement(svg, 'g')` and `svg.get("width").rstrip("mm")`

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,168 @@
+			inkex.errormsg(gettext("[INFO]: regmark to regmark spacing X ")+str(reg_width))
+			inkex.errormsg(gettext("[INFO]: regmark to regmark spacing Y ")+str(reg_length))
+
+		# Create a new layer.
```
you may call `inkex.Group(...)`

see https://inkscape.gitlab.io/extensions/documentation/source/inkex.elements._groups.html#inkex.elements._groups.Group

mofosyne replied:
Tried replacing that line with `inkex.Group(svg)`. Did not work. So not sure I understand that function.

t0b3 replied:
I didn't catch first, but for layers you may even use `Layer.new(label, *children, **attrs)` 

with the syntax documented here
https://inkscape.gitlab.io/extensions/documentation/source/inkex.elements._groups.html#inkex.elements._groups.Group.new

Thus, you may use one single line
```
layer = inkex.Layer.new(self.layername)

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,168 @@
+		reg_length = self.options.reglength if self.options.reglength else int(svg.get("height").rstrip("mm")) - reg_origin_Y*2
+
+		if self.options.verbose == True:
+			inkex.errormsg(gettext("[INFO]: page width ")+str(svg.get("width").rstrip("mm")))
```
curious: would this also work with `inkex.msg()` ?

https://inkscape.gitlab.io/extensions/documentation/source/inkex.base.html?highlight=msg#inkex.base.InkscapeExtension.msg

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,168 @@
+		# Create group for top right corner
+		topRight = etree.Element('{%s}g' % SVG_URI)
+		topRight.set('id', 'TopRight')
+		topRight.set('style', 'fill: black;')
```
same here

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,168 @@
+		# Create group for top right corner
+		bottomLeft = etree.Element('{%s}g' % SVG_URI)
+		bottomLeft.set('id', 'BottomLeft')
+		bottomLeft.set('style', 'fill: black;')
```
same here

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,168 @@
+		# Create group for top left corner keepout
+		topLeftKeepout = etree.Element('{%s}g' % SVG_URI)
+		topLeftKeepout.set('id', 'TopLeftKeepout')
+		topLeftKeepout.set('style', 'fill: black;')
```
same here

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,168 @@
+		# Create group for top right corner keepout
+		topRightKeepout = etree.Element('{%s}g' % SVG_URI)
+		topRightKeepout.set('id', 'TopRightKeepout')
+		topRightKeepout.set('style', 'fill: black;')
```
same here

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,168 @@
+		# Create group for bottom right corner keepout
+		bottomRightKeepout = etree.Element('{%s}g' % SVG_URI)
+		bottomRightKeepout.set('id', 'BottomRightKeepout')
+		bottomRightKeepout.set('style', 'fill: black;')
```
same here

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,168 @@
+"""
+
+import inkex
+from lxml import etree
```
`inkex` provides all the drawing, so lxml won't be needed.

mofosyne replied:
will remove when all etree referenced has been fixed and removed

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,168 @@
+from lxml import etree
+from gettext import gettext
+
+SVG_URI = u'http://www.w3.org/2000/svg'
```
thus you may drop raw xml like this

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,168 @@
+		layer.append(bottomRightKeepout)
+
+		#Lock layer
+		layer.set(inkex.addNS('insensitive', 'sodipodi'), 'true') 
```
use `set_sensitive()`

https://inkscape.gitlab.io/extensions/documentation/source/inkex.elements._base.html?highlight=set_sensitive#inkex.elements._base.BaseElement.set_sensitive

mofosyne committed:
render regmark: code review revision of inx file

mofosyne committed:
render regmark: rect and line switched to inkex

mofosyne committed:
render regmark: inkex.Group() replaces etree.Element('{%s}g' % SVG_URI)

mofosyne committed:
render regmark: inkex.utils.debug replaces inkex.errormsg

mofosyne committed:
render regmark: layer.set_sensitive(False) replaces layer.set(inkex.addNS('insensitive', 'sodipodi'), 'tru

mofosyne committed:
pack this section into it's own function add_arguments()

mofosyne committed:
inkex.base.InkscapeExtension.msg replace inkex.utils.debug

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -23,31 +23,26 @@
 		x, y = [pos * self.svg.unittouu('1mm') for pos in pos  ]
 		w, h = [pos * self.svg.unittouu('1mm') for pos in size ]
-		rect = etree.Element('{%s}rect' % SVG_URI)
+		rect = inkex.Rectangle()
```
you can use the args in the function call 
i.e. 
```
rect = Rectangle.new(x, y, w, h, id=name, style='fill: black;')
return rect
```

as documented in https://inkscape.gitlab.io/extensions/documentation/source/inkex.elements._polygons.html#inkex.elements._polygons.Rectangle.new

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -60,7 +55,7 @@ def drawRect(self, size, pos, name):
 		x1, y1, = [pos * self.svg.unittouu('1mm') for pos in posStart]
 		x2, y2, = [pos * self.svg.unittouu('1mm') for pos in posEnd  ]
-		line = etree.Element('{%s}line' % SVG_URI)
+		line = inkex.Line()
```
you can use i.e. 
```
line = inkex.Line.new((x1, y1), (x2, y2), id=name, style='stroke....')
return line
```

see also https://inkscape.gitlab.io/extensions/documentation/source/inkex.elements._polygons.html#inkex.elements._polygons.Line

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -102,7 +97,7 @@ def effect(self):
 		
 		# Create group for top right corner
-		topRight = etree.Element('{%s}g' % SVG_URI)
+		topRight = inkex.Group()
```
you may use `Group.new(label, *children, **attrs)`

thus i.e.
```
topRight = inkex.Group.new(id = 'TopRight', style = 'fill: black;')
```

with the syntax documented here
https://inkscape.gitlab.io/extensions/documentation/source/inkex.elements._groups.html#inkex.elements._groups.Group.new

mofosyne committed:
render regmark: compact inkex.Group and add new layer to self.svg

mofosyne committed:
render regmark: add style param  to inkex.Rectangle.new()

mofosyne committed:
inkex.Line.new() param style added. Also consolidate consts

mofosyne committed:
render regmark: remove lxml reference

mofosyne committed:
render regmark: mm_to_user_unit instead of self.svg.unittouu('1mm') for code clarity

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+		layer = self.svg.add(inkex.Layer.new(self.LAYERNAME))
+	
+		# Create square in top left corner
+		layer.append(self.drawRect((self.REG_SQUARE_MM,self.REG_SQUARE_MM), (reg_origin_X,reg_origin_Y), 'TopLeft'))
```
so this can now be simplified to 
```
layer.append(inkex.Rectangle.new(self.REG_SQUARE_MM, self.REG_SQUARE_MM, reg_origin_X, reg_origin_Y, id='TopLeft', style='fill: black;'))
```
... and `def drawRect()` can be dropped :)

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+		# Create horizontal and vertical lines in group
+		top_right_reg_origin_x = reg_origin_X+reg_width
+		topRight.append(self.drawLine((top_right_reg_origin_x-self.REG_LINE_MM,reg_origin_Y), (top_right_reg_origin_x,reg_origin_Y), 'Horizontal'))
+		topRight.append(self.drawLine((top_right_reg_origin_x,reg_origin_Y), (top_right_reg_origin_x,reg_origin_Y + self.REG_LINE_MM), 'Vertical'))
```
... and this can be simplified to 
```
top_right_reg_origin_x = reg_origin_X+reg_width
topRight = inkex.PathElement.new([(top_right_reg_origin_x-self.REG_LINE_MM,reg_origin_Y), (top_right_reg_origin_x,reg_origin_Y), (top_right_reg_origin_x,reg_origin_Y + self.REG_LINE_MM)], id = 'TopRight', style = self.LINE_STYLE)

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+		topRight.append(self.drawLine((top_right_reg_origin_x,reg_origin_Y), (top_right_reg_origin_x,reg_origin_Y + self.REG_LINE_MM), 'Vertical'))
+		layer.append(topRight)
+		
+		# Create group for top right corner
```
probably `bottom left corner`

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+		# Create horizontal and vertical lines in group
+		top_right_reg_origin_y = reg_origin_Y+reg_length
+		bottomLeft.append(self.drawLine((reg_origin_X,top_right_reg_origin_y), (reg_origin_X+self.REG_LINE_MM,top_right_reg_origin_y), 'Horizontal'))
+		bottomLeft.append(self.drawLine((reg_origin_X,top_right_reg_origin_y), (reg_origin_X,top_right_reg_origin_y - self.REG_LINE_MM), 'Vertical'))
```
... same here: PathElement()

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+		top_left_keepout_origin_x = reg_origin_X+self.REG_LINE_MM
+		top_left_keepout_origin_y = reg_origin_Y+self.REG_LINE_MM
+		topLeftKeepout.append(self.drawLine((top_left_keepout_origin_x,top_left_keepout_origin_y), (top_left_keepout_origin_x-self.REG_KEEPOUT_MM,top_left_keepout_origin_y),'Horizontal'))
+		topLeftKeepout.append(self.drawLine((top_left_keepout_origin_x,top_left_keepout_origin_y), (top_left_keepout_origin_x,top_left_keepout_origin_y-self.REG_KEEPOUT_MM), 'Vertical'))
```
same here

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+		top_left_keepout_origin_x = reg_origin_X+reg_width-self.REG_LINE_MM
+		top_left_keepout_origin_y = reg_origin_Y+self.REG_LINE_MM
+		topRightKeepout.append(self.drawLine((top_left_keepout_origin_x,top_left_keepout_origin_y), (top_left_keepout_origin_x+self.REG_KEEPOUT_MM,top_left_keepout_origin_y),'Horizontal'))
+		topRightKeepout.append(self.drawLine((top_left_keepout_origin_x,top_left_keepout_origin_y), (top_left_keepout_origin_x,top_left_keepout_origin_y-self.REG_KEEPOUT_MM), 'Vertical'))
```
same here

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+		bottom_right_keepout_origin_x = reg_origin_X+self.REG_LINE_MM
+		bottom_right_keepout_origin_y = reg_origin_Y+reg_length-self.REG_LINE_MM
+		bottomRightKeepout.append(self.drawLine((bottom_right_keepout_origin_x,bottom_right_keepout_origin_y), (bottom_right_keepout_origin_x-self.REG_KEEPOUT_MM,bottom_right_keepout_origin_y),'Horizontal'))
+		bottomRightKeepout.append(self.drawLine((bottom_right_keepout_origin_x,bottom_right_keepout_origin_y), (bottom_right_keepout_origin_x,bottom_right_keepout_origin_y+self.REG_KEEPOUT_MM), 'Vertical'))
```
same here

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+		# > The thicker you get, the less accurate registration will be. ~~~ galaxyman47
+		self.REG_MARK_LINE_WIDTH_MM = 0.3
+
+		self.mm_to_user_unit = self.svg.unittouu('1mm')
```
curiously asking what was the issue with `unit_to_viewport`?

mofosyne replied:
Scaling goes out of wack, gets shrunken if document scaling is different. Will try and read your suggestion and see if there is a proper usage for it.

> you may have a look at https://inkscape.gitlab.io/extensions/documentation/authors/units.html
> unit transforms and coordinate systems are well introduced there

mofosyne replied:
```python
	#SVG rect element generation routine
	def drawRect(self, size, pos, name):
		x, y = [self.svg.unit_to_viewport(pos, "mm") for pos in pos  ]
		w, h = [self.svg.unit_to_viewport(pos, "mm") for pos in size ]
		return Rectangle.new(x, y, w, h, id=name, style='fill: black;')
		
	#SVG line element generation routine
	def drawLine(self, posStart, posEnd, name):
		x1, y1, = [self.svg.unit_to_viewport(pos, "mm") for pos in posStart]
		x2, y2, = [self.svg.unit_to_viewport(pos, "mm") for pos in posEnd  ]
		line_style = 'stroke: black; stroke-width: '+str(self.svg.unit_to_viewport(REG_MARK_LINE_WIDTH_MM, "mm"))+';'
		return Line.new((x1, y1), (x2, y2), id=name, style=line_style)
```

Attempted with the above, but got

![image](https://github.com/fablabnbg/inkscape-silhouette/assets/827793/63f61c9a-cbf4-45b3-be9d-ae5267a7d48c)

When trying to apply this render to the existing regmark template for verification (which has different viewport). Spent a bit of trying to figure it out without any luck.

I wonder if this bit is the cause...

> Obviously the element needs to know the viewport of its SVG document for this. This method therefore does not work if the element is unrooted.

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+		# https://www.reddit.com/r/silhouettecutters/comments/wcdnzy/the_key_to_print_and_cut_success_an_extensive/
+		# > The registration mark thickness is actually very important. For some reason, 0.3 mm marks work perfectly. 
+		# > The thicker you get, the less accurate registration will be. ~~~ galaxyman47
+		self.REG_MARK_LINE_WIDTH_MM = 0.3
```
it would be simpler to introduce those CONSTANTS w/o self - i.e. before Line 25

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+Base module for rendering regmarks for Silhouette CAMEO products in Inkscape.
+"""
+
+import inkex
```
for simpler readability within the code you may use
```
from inkex import Boolean, PathElement, Rectangle # etc.
```
... and then call in the code simply `Rectangle(...)`

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+		bottomLeft.append(self.drawLine((reg_origin_X,top_right_reg_origin_y), (reg_origin_X,top_right_reg_origin_y - self.REG_LINE_MM), 'Vertical'))
+		layer.append(bottomLeft)
+		
+		# Keepout Marker #
```
I didn't get the intention of those keepout markers. Are the needed for silhouette or intended as visual marks, but what exactly would they mark?

mofosyne replied:
This keepout marker is not required for silhouette alignment function, instead it's an indicator for the user to not put any design within the keepout that would interfere with the scanning function of the machine.

The system works fast enough that it's hard to find the minimum bounding box, but it appears that it checks from the middle of each box then moves towards the line. So in theory we could reduce the keepout area...

But I'm sticking to the keepout area that is defined in your original regmark template. The way you did it in your original template was enable gridlines as the background, but then added a white fill lines on the area you can draw on. Seems like a lot of work to emulate that same feature in code, so stuck to keeping it simple with a visual keepout marker instead. Hope that explains my approach.

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+		layer.append(topRight)
+		
+		# Create group for top right corner
+		bottomLeft = inkex.Group(id = 'BottomLeft', style = 'fill: black;')
```
NB: fill on lines won't do nothing

mofosyne committed:
render regmark: constants don't need self. specifiers if moved to top

mofosyne committed:
render regmark: clean up code for inkex calls

mofosyne committed:
render regmark: inkex imports Layer and Group

mofosyne committed:
render regmark: Group() doesn't need style param. Code cleanup

mofosyne committed:
render regmark: update comment on keepout marker

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -75,54 +76,56 @@ def effect(self):
 		
 		# Keepout Marker #
+		# Not directly part of Silhouette registration marker
+		# Instead it's a visual indicator to the user to avoid putting any design within this area
```
thanks for this explanation on providing visual guidance to the user. 
would you agree to switch to something softer i.e.  similar to what is used in 
the preexisting template here https://github.com/fablabnbg/inkscape-silhouette/blob/main/templates/silhouette-cameo-registration-marks-a4.svg
* checkerboard: true https://github.com/fablabnbg/inkscape-silhouette/blob/af3c9019e948aa56c9a357263eec45a35df18f20/templates/silhouette-cameo-registration-marks-a4.svg#L42
* path with fill: `#ffffff` https://github.com/fablabnbg/inkscape-silhouette/blob/af3c9019e948aa56c9a357263eec45a35df18f20/templates/silhouette-cameo-registration-marks-a4.svg#L137-L142

mofosyne replied:
I've gave your proposal a thought, but I'm unconvinced on the marginal utility vs effort of adding this in. Plus I'm a bit hesitant of a "render" function doing pagewide settings changes.

If you can't print in that area, then there is no point in adding "softer" cutout, in fact you may mislead users in thinking that it's a usable area. I think it's better to just make it clear that you can't use that area via a simple line.

Admittedly, I'm looking at this from a utilitarian perspective and yes it's clear that it's a little bit aesthetically uglier hence why silhouette studio probably opted for that style.

Anyway, I think this is something to revisit in the future if needed as a separate PR. This is good enough in my opinion as a minimum viable pull request (in terms of code complexity).

mofosyne replied:
Could I make a separate issue ticket to revisit with your proposal and tips on implementing it? 

E.g. "Feature Enhancement: Render Regmark to have softer keepout area"

mofosyne replied:
Okay I've thought it though a bit more.

A safe area keepout is a good idea, as I've seen online people using that keepout area as just extra whitespace when trimming.

I've added it in now.

![image](https://github.com/fablabnbg/inkscape-silhouette/assets/827793/bcf9e3a4-2171-422b-a42f-1bd70f7bcbb7)

Also keeping to 20mm x 20mm keepout area as it's possible that it's needed when angling the paper.

As for `inkscape:pagecheckerboard="true"`, looks like dark mode inkscape at least does not have that issue. Plus as mentioned above, may not be good practice to unexpectedly set page wide settings.

Kept safe area as a separate layer, as people may want to adjust the color of it or just delete it.

mofosyne committed:
render regmark: Add print safe area layer instead of keepout lines

mofosyne committed:
render regmark: Adjust naming of print safe area

mofosyne committed:
render regmark: use from inkex.extensions import EffectExtension

mofosyne committed:
render regmark: fix safe area scaling. Add mark notes

mofosyne committed:
render regmark: Fix text scaling

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -74,6 +64,17 @@ def drawLine(self, posStart, posEnd, name):
+			svgd += " L{:.5f},{:.5f}".format(x[0]*mm_to_user_unit, x[1]*mm_to_user_unit)
+		svgd += "z"
+		return svgd
+
```
you may use `Polygon.new(points=p)` with a list of points instead
NB: of course append `Polygon` to the inkex imports :)

mofosyne replied:
Attempted

```python
safearea.set_path(self.points_to_svgd(points))
-->
safearea.set_path(Polygon.new(points=points))
```

Got

```
/usr/share/inkscape/extensions/inkex/paths.py:1350: FutureWarning: The behavior of this method will change in future versions. Use specific 'len(elem)' or 'elem is not None' test instead.
  for item in path_d or ():
```

And no polygon is shown.

mofosyne committed:
render regmark: Fix comments for bottom left corner

mofosyne committed:
render regmark: cleanup for clarity

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,157 @@
+import inkex
+from inkex.extensions import EffectExtension
+from inkex import Boolean, Rectangle, Line, PathElement
+from inkex import Layer, Group, TextElement
```
to simplify replace this with (depending on which classes you really use)
```
from inkex import EffectExtension, Boolean, Rectangle, Polygon, PolyLine, PathElement, Layer, Group, TextElement

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,157 @@
+		for x in p:
+			svgd += " L{:.5f},{:.5f}".format(x[0]*mm_to_user_unit, x[1]*mm_to_user_unit)
+		svgd += "z"
+		return svgd
```
to simplify drop this entirely (you will see why later on :smile: )

mofosyne replied:
was not able to drop it, because Polygon() approach hasn't worked out yet.

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,157 @@
+			inkex.base.InkscapeExtension.msg(gettext("[INFO]: regmark from document left ")+str(reg_origin_X))
+			inkex.base.InkscapeExtension.msg(gettext("[INFO]: regmark from document top ")+str(reg_origin_Y))
+			inkex.base.InkscapeExtension.msg(gettext("[INFO]: regmark to regmark spacing X ")+str(reg_width))
+			inkex.base.InkscapeExtension.msg(gettext("[INFO]: regmark to regmark spacing Y ")+str(reg_length))
```
the function you want to call is `self.msg(`

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,157 @@
+		regmark_layer = self.svg.add(Layer.new(REGMARK_LAYERNAME))
+	
+		# Create square in top left corner
+		regmark_layer.append(self.drawRect((REG_SQUARE_MM,REG_SQUARE_MM), (reg_origin_X,reg_origin_Y), 'TopLeft'))
```
call `Rectangle(` directly, so you don't need custom function `drawRect`
```
regmark_layer.append(Rectangle((REG_SQUARE_MM,REG_SQUARE_MM), (reg_origin_X,reg_origin_Y), id='TopLeft'), style='fill: black;')

mofosyne replied:
drawRect() dropped

mofosyne committed:
render regmark: cleanup of points_to_svgd()

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,157 @@
+		top_right_reg_origin_x = reg_origin_X+reg_width
+		topRight.append(self.drawLine((top_right_reg_origin_x-REG_LINE_MM,reg_origin_Y), (top_right_reg_origin_x,reg_origin_Y), 'Horizontal'))
+		topRight.append(self.drawLine((top_right_reg_origin_x,reg_origin_Y), (top_right_reg_origin_x,reg_origin_Y + REG_LINE_MM), 'Vertical'))
+		regmark_layer.append(topRight)
```
simplify this by calling `PathElement.new(` directly w/o custom function `drawLine(`
```
top_right_reg_origin_x = reg_origin_X+reg_width
top_right_path = [(top_right_reg_origin_x-REG_LINE_MM,reg_origin_Y), (top_right_reg_origin_x,reg_origin_Y), (top_right_reg_origin_x,reg_origin_Y + REG_LINE_MM)]
regmark_layer.append(PathElement.new(path=top_right_path, id="TopRight", style=f"stroke: black; stroke-width: {REG_MARK_LINE_WIDTH_MM}"))

mofosyne replied:
drawLine() dropped now

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,157 @@
+		top_right_reg_origin_y = reg_origin_Y+reg_length
+		bottomLeft.append(self.drawLine((reg_origin_X,top_right_reg_origin_y), (reg_origin_X+REG_LINE_MM,top_right_reg_origin_y), 'Horizontal'))
+		bottomLeft.append(self.drawLine((reg_origin_X,top_right_reg_origin_y), (reg_origin_X,top_right_reg_origin_y - REG_LINE_MM), 'Vertical'))
+		regmark_layer.append(bottomLeft)
```
same here: simplify this by calling `PathElement.new(` directly w/o custom function `drawLine(`
```
bottom_left_reg_origin_y = reg_origin_Y+reg_length
bottom_left_path = [...]
regmark_layer.append(PathElement.new(path=bottom_left_path, id="BottomLeft", style=f"stroke: black; stroke-width: {REG_MARK_LINE_WIDTH_MM}"))
```
PS: there seems to be some confusion of top_right and bottom_left (probably copy&pasted)

mofosyne replied:
yup it likely was. Fixed

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,156 @@
+			(top_right_safearea_origin_x,top_right_safearea_origin_y-REG_SAFE_AREA_MM),
+			(top_right_safearea_origin_x,top_right_safearea_origin_y),
+			(top_right_safearea_origin_x+REG_SAFE_AREA_MM,top_right_safearea_origin_y),
+			(top_right_safearea_origin_x+REG_SAFE_AREA_MM,bottom_right_safearea_origin_y+REG_SAFE_AREA_MM),
```
probably `bottom_right?` ... but check with the output

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,156 @@
+			(top_right_safearea_origin_x+REG_SAFE_AREA_MM,bottom_right_safearea_origin_y+REG_SAFE_AREA_MM),
+			(bottom_right_safearea_origin_x,bottom_right_safearea_origin_y+REG_SAFE_AREA_MM),
+			(bottom_right_safearea_origin_x,bottom_right_safearea_origin_y),
+			(bottom_right_safearea_origin_x-REG_SAFE_AREA_MM,bottom_right_safearea_origin_y),
```
probably `bottom_left`? (namiing confusion :smile: )

mofosyne replied:
aye correct. naming confusion. Looked over, hopefully fixed now

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,156 @@
+		]
+		safearea_polygon = PathElement(id="SafeArea", style='display:inline;fill:#ffffff;stroke:none;stroke-dasharray:1, 1')
+		safearea_polygon.set_path(self.points_to_svgd(safe_area_points))
+		regmark_layer.append(safearea_polygon)
```
... hence you can siplify to
```
regmark_layer.append(Polygon.new(points=safearea_polygon, id="SafeArea", style='display:inline;fill:#ffffff;stroke:none;stroke-dasharray:1, 1')

mofosyne replied:
Attempted to use Polygon without success... is there a difference in how it handles style or something? Also still have to use points_to_svgd()

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,156 @@
+		safeare_notes_text_element.set('x', (top_left_safearea_origin_x+3) * self.svg.unittouu('1mm'))
+		safeare_notes_text_element.set('y', (bottom_right_safearea_origin_y+(REG_SAFE_AREA_MM+reg_origin_Y/2))*self.svg.unittouu('1mm'))
+		safeare_notes_text_element.set('font-size', 2.5 * self.svg.unittouu('1mm'))
+		regmark_layer.append(safeare_notes_text_element)
```
... and here continue the simplification
```
regmark_layer.append(TextElement.new(safe_area_note, id="RegMarkNotes", x= ..., y=..., style="font-size: 2.5"))

mofosyne replied:
simplified text completed

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,156 @@
+		regmark_layer.append(safeare_notes_text_element)
+
+		# Lock Layer
+		regmark_layer.set_sensitive(False)
```
you meant maybe another layer? :smile:

mofosyne replied:
fixed

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,156 @@
+
+		# Create a new register mark layer
+		regmark_layer = self.svg.add(Layer.new(REGMARK_LAYERNAME))
+	
```
treat scaling here once for all
```
regmark_layer.transform=Transform(...

mofosyne replied:
transform fixed. Interestingly and annoyingly these functions plus TextElement() uses textual input... why... very annoying.

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,156 @@
+		
+		# Create horizontal and vertical lines in group for top right corner
+		topRight = Group(id = 'TopRight')
+		top_right_reg_origin_x = reg_origin_X+reg_width
```
name top_right_reg_origin_x is difficult to understand: what about top_right_x?

mofosyne replied:
yup, you are right. that naming is easier to read

mofosyne committed:
render regmark: cleanup from import and fix self.msg

mofosyne committed:
render regmark: code review. Use PathElement and Rectangle directly.

mofosyne committed:
render regmark: text render cleanup

mofosyne committed:
render regmark: use Transform() to scale mm to doc units

mofosyne committed:
render regmark: add REG_MARK_INFO_FONT_SIZE_PT

mofosyne committed:
render regmark: naming cleanup

mofosyne committed:
render regmark: clean up style to remove redundant css values

mofosyne committed:
render regmark: consistent naming with existing template svg

mofosyne committed:
render regmark: add REGMARK_NOTES_ID=regmark-notes

mofosyne committed:
render regmark: safe area polygon if user later wants to add lines around, should have it's path capped

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+"""
+
+import inkex
+from inkex import EffectExtension, Boolean, Rectangle, Line, PathElement, Layer, Group, TextElement, Polygon, Transform
```
drop unused: Line, Polygon

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+
+class InsertRegmark(EffectExtension):
+	def __init__(self):
+		EffectExtension.__init__(self)
```
you only need to redefine `__init__` if you add more than just calling parent __init__
thus you can drop this

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+			svgd += " L{:.5f},{:.5f}".format(x[0], x[1])
+		if capped:
+			svgd += "z"
+		return svgd
```
you may drop this part entirely

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+		# Create horizontal and vertical lines in group for top right corner
+		top_right_x = reg_origin_X+reg_width
+		top_right_path = [(top_right_x-REG_LINE_MM,reg_origin_Y), (top_right_x,reg_origin_Y), (top_right_x,reg_origin_Y + REG_LINE_MM)]
+		regmark_layer.append(PathElement.new(path=self.points_to_svgd(top_right_path), id=REGMARK_TOP_RIGHT_ID, style=f"fill:none; stroke:black; stroke-width:{REG_MARK_LINE_WIDTH_MM}"))
```
you may reformulate:
```
regmark_layer.append(PathElement.new(path="M"+str(top_right_path), id=REGMARK_TOP_RIGHT_ID, style=f"fill:none; stroke:black; stroke-width:{REG_MARK_LINE_WIDTH_MM}"))

mofosyne replied:
How does this work? I thought `"M"+str([(1,2),(3,4)])` is `M[(1, 2), (3, 4)]`

t0b3 replied:
... which is then parsed successfully by the string parser: it just works TM :smile: HTH

to be honest: I would have liked to pass a list directly, which generates wrong svg due to a bug uncovered.
and thanks to your explanation on failing `Polygon.new(points=list_of_points)` - you uncovered yet another issue in inkex :+1:

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+		# Create horizontal and vertical lines in group for bottom left corner
+		bottom_left_y = reg_origin_Y+reg_length
+		bottom_left_path = [(reg_origin_X+REG_LINE_MM,bottom_left_y), (reg_origin_X,bottom_left_y), (reg_origin_X,bottom_left_y - REG_LINE_MM)]
+		regmark_layer.append(PathElement.new(path=self.points_to_svgd(bottom_left_path), id=REGMARK_BOTTOM_LEFT_ID, style=f"fill:none; stroke:black; stroke-width:{REG_MARK_LINE_WIDTH_MM}"))
```
you can replace with:
```
regmark_layer.append(PathElement.new(path="M"+str(bottom_left_path), id=REGMARK_BOTTOM_LEFT_ID, style=f"fill:none; stroke:black; stroke-width:{REG_MARK_LINE_WIDTH_MM}"))

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+			(bottom_left_safearea_origin_x,bottom_left_safearea_origin_y),
+			(bottom_left_safearea_origin_x-REG_SAFE_AREA_MM,bottom_left_safearea_origin_y),
+		]
+		regmark_layer.append(PathElement.new(path=self.points_to_svgd(safe_area_points, capped=True), id=REGMARK_SAFE_AREA_ID, style='fill:white;stroke:none'))
```
you can use
```
regmark_layer.append(PathElement.new(path="M"+str(safe_area_points)+"z", id=REGMARK_SAFE_AREA_ID, style='fill:white;stroke:none'))

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+
+		# Add some settings reminders to the print layer as a reminder
+		safe_area_note = f"mark distance from document: Left={reg_origin_X}mm, Top={reg_origin_Y}mm; mark to mark distance: X={reg_width}mm, Y={reg_length}mm; "
+		regmark_layer.append(TextElement(safe_area_note, x=f"{(bottom_left_safearea_origin_x+3)}", y=f"{(bottom_left_safearea_origin_y+(REG_SAFE_AREA_MM+reg_origin_Y/2))}", id = REGMARK_NOTES_ID, style=f"font-size:{REG_MARK_INFO_FONT_SIZE_PT}px"))
```
WRT `font-size:{REG_MARK_INFO_FONT_SIZE_PT}px` choose either `pt` or `px` for both: name and unit :smile:

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+
+		# Create a new register mark layer
+		regmark_layer = self.svg.add(Layer.new(REGMARK_LAYERNAME, id=REGMARK_LAYER_ID))
+		regmark_layer.transform = Transform(f"scale({mm_to_user_unit}, {mm_to_user_unit})")
```
- you only need to give 1 scale arg for `Transform(scale = 2)` if scaling x,y is equal
- did you test scaling on different page sizes, page units and docscales?

mofosyne replied:
Argh... you are right... epic fail. `self.svg.get("height").rstrip("mm")` breaks if in American page size. Gonna have to revisit how units works... and how the other suggested alternative to unittouu() actually works -.-

mofosyne replied:
Okay looks like `self.svg.to_dimensional(self.svg.get("width"), "mm")` is able to handle such american sized docs.

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+		mm_to_user_unit = self.svg.unittouu('1mm')
+
+		# Create a new register mark layer
+		regmark_layer = self.svg.add(Layer.new(REGMARK_LAYERNAME, id=REGMARK_LAYER_ID))
```
- currently the safearea covers anything else, so you won't see or print anything.
  - you may choose `.insert` over `.add` and choose a default value for smart layer sorting
- and the safearealayer isn't distinguishable except if you manually set `pagecheckerboard=true`
  - so you can automate and set pagecheckerboard with this plugin for a nice UX

mofosyne replied:
Yup `.insert(0, ... )` is the approach here.
Also `self.svg.namedview.set(inkex.addNS('pagecheckerboard', 'inkscape'), str(ENABLE_CHECKERBOARD).lower())` is how to set the checkerboard pattern. Just found `self.svg.namedview` when looking at existing code examples in inkscape default extensions.

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,132 @@
+		reg_origin_X = self.options.regoriginx
+		reg_origin_Y = self.options.regoriginy
+		reg_width = self.options.regwidth if self.options.regwidth else int(self.svg.get("width").rstrip("mm")) - reg_origin_X*2
+		reg_length = self.options.reglength if self.options.reglength else int(self.svg.get("height").rstrip("mm")) - reg_origin_Y*2
```
not all the documents are in 'mm'. there might be documents in inch and other units as well... :smile:

mofosyne replied:
Solved via `self.svg.to_dimensional(self.svg.get("width"), "mm")`

mofosyne committed:
render regmark: points_to_svgd() removed. remove redundant __init__

mofosyne committed:
render regmark: insert regmark layer to lowest svg stack position

mofosyne committed:
render regmark: import clean up. Removed unused Line and Polygon reference

mofosyne committed:
render regmark: enable checkboard pattern. Captialise Z in path string

mofosyne committed:
render regmark: simplify Transform() call in regmark layer rescaling

mofosyne committed:
render regmark: use self.svg.to_dimensional() to deal with documents defined in inches

mofosyne committed:
render regmark: delete any existing regmark layer

mofosyne committed:
render regmark: use const REGMARK_LAYER_ID for deleting existing regmark layer

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,131 @@
+		pars.add_argument("-X", "--reg-x", "--regwidth",  type = float, dest = "regwidth",   default = 180.0, help="X mark distance [mm]")
+		pars.add_argument("-Y", "--reg-y", "--reglength", type = float, dest = "reglength",  default = 230.0, help="Y mark distance [mm]")
+		pars.add_argument("--rego-x",  "--regoriginx",    type = float, dest = "regoriginx", default = 15.0,  help="X mark origin from left [mm]")
+		pars.add_argument("--rego-y", "--regoriginy",     type = float, dest = "regoriginy", default = 20.0,  help="X mark origin from top [mm]")
```
wanna sync default values with INX? (0,0,10,10)
or is there a good reason to differ from it?

mofosyne replied:
Yup we should sync. Fixed

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,131 @@
+	def effect(self):
+		reg_origin_X = self.options.regoriginx
+		reg_origin_Y = self.options.regoriginy
+		reg_width = self.options.regwidth if self.options.regwidth else self.svg.to_dimensional(self.svg.get("width"), "mm") - reg_origin_X*2
```
reg_width = self.options.regwidth if self.options.regwidth else self.svg.to_dimensional(self.svg.viewport_width, "mm") - reg_origin_X * 2

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,131 @@
+		reg_origin_X = self.options.regoriginx
+		reg_origin_Y = self.options.regoriginy
+		reg_width = self.options.regwidth if self.options.regwidth else self.svg.to_dimensional(self.svg.get("width"), "mm") - reg_origin_X*2
+		reg_length = self.options.reglength if self.options.reglength else self.svg.to_dimensional(self.svg.get("height"), "mm") - reg_origin_Y*2
```
reg_length = self.options.reglength if self.options.reglength else self.svg.to_dimensional(self.svg.viewport_height, "mm") - reg_origin_Y * 2

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,131 @@
+		reg_length = self.options.reglength if self.options.reglength else self.svg.to_dimensional(self.svg.get("height"), "mm") - reg_origin_Y*2
+
+		if self.options.verbose == True:
+			self.msg(gettext("[INFO]: page width ")+str(self.svg.to_dimensional(self.svg.get("width"), "mm")))
```
self.msg(gettext("[INFO]: page width ")+str(self.svg.to_dimensional(self.svg.viewport_width, "mm")))

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,131 @@
+
+		if self.options.verbose == True:
+			self.msg(gettext("[INFO]: page width ")+str(self.svg.to_dimensional(self.svg.get("width"), "mm")))
+			self.msg(gettext("[INFO]: page height ")+str(self.svg.to_dimensional(self.svg.get("height"), "mm")))
```
self.msg(gettext("[INFO]: page height ")+str(self.svg.to_dimensional(self.svg.viewport_height, "mm")))

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,131 @@
+
+		# Check if existing regmark layer exist and delete it
+		old_regmark_layer = self.svg.getElementById(REGMARK_LAYER_ID)
+		if old_regmark_layer is not None:
```
if old_regmark_layer:

mofosyne replied:
Attempted to remove, but got this warning message.

```
/usr/share/inkscape/extensions/render_silhouette_regmarks.py:72: FutureWarning: The behavior of this method will change in future versions. Use specific 'len(elem)' or 'elem is not None' test instead.
  if old_regmark_layer:
```

So placing it back in.

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,131 @@
+			old_regmark_layer.delete()
+
+		# Register Mark #
+		mm_to_user_unit = self.svg.unittouu('1mm')
```
mm_to_user_unit = self.svg.viewport_to_unit("1mm")

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,131 @@
+		# Create horizontal and vertical lines in group for top right corner
+		top_right_x = reg_origin_X+reg_width
+		top_right_path = [(top_right_x-REG_LINE_MM,reg_origin_Y), (top_right_x,reg_origin_Y), (top_right_x,reg_origin_Y + REG_LINE_MM)]
+		regmark_layer.append(PathElement.new(path="M"+str(top_right_path), id=REGMARK_TOP_RIGHT_ID, style=f"fill:none; stroke:black; stroke-width:{REG_MARK_LINE_WIDTH_MM}"))
```
trailing `;` in style attr

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,131 @@
+		# Create horizontal and vertical lines in group for bottom left corner
+		bottom_left_y = reg_origin_Y+reg_length
+		bottom_left_path = [(reg_origin_X+REG_LINE_MM,bottom_left_y), (reg_origin_X,bottom_left_y), (reg_origin_X,bottom_left_y - REG_LINE_MM)]
+		regmark_layer.append(PathElement.new(path="M"+str(bottom_left_path), id=REGMARK_BOTTOM_LEFT_ID, style=f"fill:none; stroke:black; stroke-width:{REG_MARK_LINE_WIDTH_MM}"))
```
trailing `;` in style attr

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,131 @@
+		top_right_safearea_origin_x = reg_origin_X+reg_width-REG_LINE_MM
+		top_right_safearea_origin_y = reg_origin_Y+REG_LINE_MM
+		bottom_left_safearea_origin_x = reg_origin_X+REG_LINE_MM
+		bottom_left_safearea_origin_y = reg_origin_Y+reg_length-REG_LINE_MM
```
would be more readable i.e. like
```
safearea_left_x = reg_origin_X+REG_LINE_MM
safearea_top_y = reg_origin_Y+REG_LINE_MM
safearea_right_x = reg_origin_X+reg_width-REG_LINE_MM
safearea_bottom_y = reg_origin_Y+reg_length-REG_LINE_MM
safe_area_points = [
   ...

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,131 @@
+			(bottom_left_safearea_origin_x,bottom_left_safearea_origin_y),
+			(bottom_left_safearea_origin_x-REG_SAFE_AREA_MM,bottom_left_safearea_origin_y),
+		]
+		regmark_layer.append(PathElement.new(path="M"+str(safe_area_points)+"Z", id=REGMARK_SAFE_AREA_ID, style='fill:white;stroke:none'))
```
trailing `;` in style attr

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,131 @@
+
+		# Add some settings reminders to the print layer as a reminder
+		safe_area_note = f"mark distance from document: Left={reg_origin_X}mm, Top={reg_origin_Y}mm; mark to mark distance: X={reg_width}mm, Y={reg_length}mm; "
+		regmark_layer.append(TextElement(safe_area_note, x=f"{(bottom_left_safearea_origin_x+3)}", y=f"{(bottom_left_safearea_origin_y+(REG_SAFE_AREA_MM+reg_origin_Y/2))}", id = REGMARK_NOTES_ID, style=f"font-size:{REG_MARK_INFO_FONT_SIZE_PX}px"))
```
trailing `;` in style attr

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,131 @@
+		self.svg.insert(0, regmark_layer)
+
+		# Set Page Setting to enable checkerboard (This is required so that safe area is easier to see)
+		self.svg.namedview.set(inkex.addNS('pagecheckerboard', 'inkscape'), str(ENABLE_CHECKERBOARD).lower())
```
self.svg.namedview.set('inkscape:pagecheckerboard', str(ENABLE_CHECKERBOARD).lower())

mofosyne committed:
render regmark: sync default values with inx (0,0,10,10)

mofosyne committed:
render regmark: cleanup for readability (code review)

mofosyne committed:
render regmark: adjust cli param adjusted

mofosyne committed:
render regmark: put 'is not None' test back because of FutureWarning compile message

/usr/share/inkscape/extensions/render_silhouette_regmarks.py:72: FutureWarning: The behavior of this method will change in future versions. Use specific 'len(elem)' or 'elem is not None' test instead.
  if old_regmark_layer:

mofosyne reviewed (commented):
Okay answered all the clean up suggestions (with one rejected suggestion because it triggered a warning flag). Let's see how this one goes!

mofosyne committed:
render regmark: add tests

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -19,10 +19,21 @@
 from inkex import EffectExtension, Boolean, Rectangle, PathElement, Layer, Group, TextElement, Transform
 from gettext import gettext
 
+# Enables stand alone mode and helps for tests #
+# We append the directory where this script lives and inkscape extension folder to sys.path
+sys.path.append(os.path.dirname(os.path.abspath(sys.argv[0])))
+sys_platform = sys.platform.lower()
+if sys_platform.startswith("win"):
+    sys.path.append(r"C:\Program Files\Inkscape\share\inkscape\extensions")
+elif sys_platform.startswith("darwin"):
+    sys.path.append("/Applications/Inkscape.app/Contents/Resources/share/inkscape/extensions")
+else:   # linux
+    sys.path.append("/usr/share/inkscape/extensions")
+
```
this comes **above** all the `import inkex` i.e. above line 21

mofosyne committed:
render regmark: compile fix

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,142 @@
+		reg_origin_X = self.options.regoriginx
+		reg_origin_Y = self.options.regoriginy
+		reg_width = self.options.regwidth if self.options.regwidth else self.svg.to_dimensional(self.svg.viewport_width, "mm") - reg_origin_X * 2
+		reg_length = self.options.reglength if self.options.reglength else self.svg.to_dimensional(self.svg.viewport_height, "mm") - reg_origin_Y * 2
```
can you switch from `... if ... else ...` to `... or ...`? this should work for 0 :)

t0b3 commented on test/test_render_silhouette_regmark.py:
```diff
@@ -0,0 +1,56 @@
+        )
+
+        """Ensure x distance"""
+        transform = self.e.svg.getElementById('regmark-tr').composed_transform()
```
drop transform

t0b3 commented on test/test_render_silhouette_regmark.py:
```diff
@@ -0,0 +1,56 @@
+        )
+
+        """Ensure y distance"""
+        transform = self.e.svg.getElementById('regmark-tr').composed_transform()
```
drop transform

t0b3 commented on test/test_render_silhouette_regmark.py:
```diff
@@ -0,0 +1,56 @@
+        self.assertEqual(
+            self.e.svg.unit_to_viewport(
+                    (self.e.svg.getElementById('regmark-bl').bounding_box(transform).y.maximum
+                    - self.e.svg.getElementById('regmark-tr').bounding_box(transform).y.minimum),
```
(self.e.svg.getElementById('regmark-bl').bounding_box(transform=True).y.maximum
                    - self.e.svg.getElementById('regmark-tr').bounding_box(transform=True).y.minimum),

t0b3 commented on test/test_render_silhouette_regmark.py:
```diff
@@ -0,0 +1,56 @@
+        self.assertEqual(
+            self.e.svg.unit_to_viewport(
+                    (self.e.svg.getElementById('regmark-tr').bounding_box(transform).x.maximum
+                    - self.e.svg.getElementById('regmark-bl').bounding_box(transform).x.minimum),
```
(self.e.svg.getElementById('regmark-tr').bounding_box(transform=True).x.maximum
                    - self.e.svg.getElementById('regmark-bl').bounding_box(transform=True).x.minimum),

mofosyne committed:
render regmark: backport monkey patches

mofosyne committed:
render regmar: test code cleanup

mofosyne committed:
render regmark: add BaseElement

mofosyne committed:
render regmark: test: drop transform

mofosyne committed:
render regmark: removed unused backport calls

mofosyne committed:
render regmark: Add viewport_to_unit backport

mofosyne committed:
render regmark: attempted backport of viewport_to_unit()

mofosyne committed:
render regmark: more attempt at adding possibly missing backport calls

mofosyne committed:
render regmark: For regmark rendering inkex < 1.2 is not supported

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,153 @@
+	SvgDocumentElement.to_dimensional = staticmethod(lambda self, value, to_unit="px": convert_unit(value, to_unit))
+	SvgDocumentElement.to_dimensionless = staticmethod(lambda self, value: convert_unit(value, "px"))
+	SvgDocumentElement.equivalent_transform_scale = property(lambda self: max([self.to_dimensional(self.viewport_width, unit) / self.viewbox_width, self.to_dimensional(self.viewport_height, unit) / self.viewbox_height]) or 1.0)
+	SvgDocumentElement.viewport_to_unit = staticmethod(lambda self, value, unit="px": self.to_dimensional(self.to_dimensionless(value) / self.root.equivalent_transform_scale, unit))
```
this error was because you've put self arg for static methods
and watch out `equivalent_transform_scale` relies on `scale` which calls `_base_scale()` which already has a sophisticated logic backported ...
```
if not hasattr(inkex, "__version__") or inkex.__version__[0:3] < "1.2":
    from inkex import BaseElement, SvgDocumentElement, paths
    import re
    SvgDocumentElement.viewport_width = property(lambda self: convert_unit(self.get("width"), "px") or self.get_viewbox()[2])
    SvgDocumentElement.viewport_height = property(lambda self: convert_unit(self.get("height"), "px") or self.get_viewbox()[3])
    SvgDocumentElement._base_scale = lambda self, unit="px": (convert_unit(1, unit) or 1.0) if not all(self.get_viewbox()[2:]) else max([convert_unit(self.viewport_width, unit) / self.get_viewbox()[2], convert_unit(self.viewport_height, unit) / self.get_viewbox()[3]]) or convert_unit(1, unit) or 1.0
    BaseElement.to_dimensional = staticmethod(lambda value, to_unit="px": convert_unit(value, to_unit))
    BaseElement.to_dimensionless = staticmethod(lambda value: convert_unit(value, "px"))
    BaseElement.viewport_to_unit = lambda self, value, unit="px": self.to_dimensional(self.to_dimensionless(value) / self.root._base_scale(), unit)
    BaseElement.unit_to_viewport = lambda self, value, unit="px": self.to_dimensional(self.to_dimensionless(value) * self.root._base_scale(), unit)
    BaseElement.set_sensitive = lambda self, sensitive="true": self.set("sodipodi:insensitive", ["true", None][sensitive])
    paths.strargs = lambda string, kind=float: [kind(val) for val in re.compile(r"(?:[+-]?(?:(?:(?:[0-9]+)?\.(?:[0-9]+)|(?:[0-9]+)\.)(?:[eE][+-]?(?:[0-9]+))?|(?:[0-9]+)(?:[eE][+-]?(?:[0-9]+)))|[+-]?(?:[0-9]+))").findall(string)]
```

PS: maybe you can profit from some of those monkeypatches for #253

t0b3 commented on render_silhouette_regmarks.py:
```diff
@@ -0,0 +1,153 @@
+    sys.path.append("/usr/share/inkscape/extensions")
+
+import inkex
+from inkex import EffectExtension, Boolean, Rectangle, PathElement, Layer, Group, TextElement, Transform, BaseElement, SvgDocumentElement
```
drop `BaseElement, SvgDocumentElement`

mofosyne committed:
render regmark: adjust monkeypatch for inkscape v1.1
